forc-pkg/src/pkg.rs

```diff
@@ -2490,16 +2518,12 @@ fn build_profile_from_opts(
     Ok((selected_build_profile.to_string(), profile))
 }
 
-/// Builds a project with given BuildOptions
-pub fn build_with_options(build_options: BuildOpts) -> Result<Built> {
-    let path = &build_options.pkg.path;
-
-    let this_dir = if let Some(ref path) = path {
-        PathBuf::from(path)
-    } else {
-        std::env::current_dir()?
-    };
-
+/// Builds a project with given BuildOptions while injecting provided items to the namepsace of
+/// relevant package's namespace.
+pub fn inject_and_build_with_options(
+    build_options: BuildOpts,
+    inject_map: HashMap<NodeIx, Vec<(String, ConfigTimeConstant)>>,
```

Comment:
Perhaps we can include the `inject_map` in the `BuildOpts` to remove the need for an extra constructor function?

Can we really use `NodeIx` safely as the key? Isn't it possible for the `NodeIx` to change if members are added/removed to/from the workspace? Curious to see what usage looks like. Is it possible we want to use `Pkg` as a key instead?

Also, perhaps we could use some type aliases to make this type a little easier to read? e.g.

```rust
type ConstName = String;
type ConstInjectionMap = HashMap<NodeIx, Vec<(ConstName, ConfigTimeConstant)>>;
```

We should really change the name from `ConfigTimeConstant` to `Constant` or something, especially now that *actual* config time constants are being introduced (#3536). Maybe it's time to prioritize #3077?

Reply:
It seems to me that `NodeIx` stays the same but using `pkg` might be better here as it would easier to understand. I planning to tackle #3077 right after #3729 👍

Reply:
I ended up using `Pinned` since I need the get a `Path` reference to convert `Pinned` to `Pkg` as `Pinned` is the struct that I can directly reach from `BuildPlan`. This feels more coherent to me as well since we are injecting stuff to the nodes of `BuildPlan`'s `graph` (which are `Pinned`s). To be able to use the new injection map easily from outside of `forc-pkg` I also added a util function that returns an iterator over `Pinned`s of member packages.

Reply:
Nice, this looks a little more fool-proof! Looking forward to seeing how this gets used in `forc-test` :ok_hand: